torch/testing/_internal/distributed/rpc/rpc_test.py

```diff
@@ -5635,3 +5660,111 @@ def test_rref_to_here_synchronization_cross_device(self):
                 self.assertEqual(actual, expected)
 
         rpc.shutdown()
+
+    @skip_if_lt_x_gpu(1)
+    def test_cuda_future_device_as_int(self):
+        fut = Future([0])
+
+    @skip_if_lt_x_gpu(1)
+    def test_cuda_future_device_as_str(self):
+        fut = Future(["cuda:0"])
+
+    @skip_if_lt_x_gpu(1)
+    def test_cuda_future_device_as_device(self):
+        fut = Future([torch.device("cuda", 0)])
+
+    def _test_cuda_future_extraction(self, wrapper, unwrapper):
+        # We check proper CUDA stream synchronization by filling the tensor with
+        # the expected value in one stream, and reading it from another stream.
+        tensor = torch.zeros((100,), device="cuda:0")
+        future = Future(["cuda:0"])
+        with torch.cuda.device("cuda:0"):
+            stream = torch.cuda.Stream()
+            another_stream = torch.cuda.Stream()
+            with torch.cuda.stream(stream):
+                torch.cuda._sleep(int(1000 * get_cycles_per_ms()))
+                tensor.fill_(1)
+                future.set_result(wrapper(tensor))
+            with torch.cuda.stream(another_stream):
+                self.assertTrue(torch.equal(
+                    unwrapper(future.wait()).cpu(), torch.ones((100,))
```

Comment:
Not an expert on cuda stream synchronization so this comment might not be relevant. But should we be doing this check on the device instead of moving to CPU? AFAIK, moving to CPU triggers a device synchronize so all streams are synchronized when `.cpu()` is called. 

What we want to test is that `event.block()` in postWaitHook is blocking the current stream appropriately right?

Reply:
Good point. I am not sure if a `.cpu()` call causes a "full" device syncs (all kernels on all streams), or just a sync for the current stream. Though we shouldn't take any chances in tests, hence I changed this to do the comparison on GPU and only copy the final boolean result back to CPU.